File: packages/swc-plugin/src/whitespace.rs
#[derive(Debug, PartialEq)]
enum WhitespaceType {
    NormalSpace,           // Regular space from typing
    Tab,                   // Tab character from indentation
    Newline,              // Newline from text formatting
    HtmlEntity,           // Decoded HTML entities like &nbsp;
    UnicodeWhitespace,    // Other Unicode whitespace
}

fn classify_whitespace_char(ch: char) -> Option<WhitespaceType> {
    match ch {
        ' ' => Some(WhitespaceType::NormalSpace),
        '\t' => Some(WhitespaceType::Tab),
        '\n' | '\r' => Some(WhitespaceType::Newline),
        '\u{00A0}' => Some(WhitespaceType::HtmlEntity), // &nbsp;
        '\u{00AD}' => Some(WhitespaceType::HtmlEntity), // &shy;
        '\u{202F}' => Some(WhitespaceType::HtmlEntity), // Narrow no-break space
        '\u{2060}' => Some(WhitespaceType::HtmlEntity), // Word joiner
        ch if ch.is_whitespace() => Some(WhitespaceType::UnicodeWhitespace),
        _ => None,
    }
}


pub fn has_significant_whitespace(text: &str) -> bool {
    for ch in text.chars() {
        match classify_whitespace_char(ch) {
            Some(WhitespaceType::NormalSpace) => continue,
            Some(WhitespaceType::Tab) => continue,
            Some(WhitespaceType::Newline) => continue,
            Some(WhitespaceType::HtmlEntity) => return true,
            Some(WhitespaceType::UnicodeWhitespace) => return true,
            None => continue, // Not whitespace
        }
    }
    false
}


fn is_normal_whitespace(ch: char) -> bool {
    match classify_whitespace_char(ch) {
        Some(WhitespaceType::NormalSpace) => true,
        Some(WhitespaceType::Tab) => true,
        Some(WhitespaceType::Newline) => true,
        Some(WhitespaceType::HtmlEntity) => false,
        Some(WhitespaceType::UnicodeWhitespace) => false,
        None => false,
    }
}

fn trim_with_callback<F>(text: &str, should_trim: F) -> &str
where
    F: Fn(char) -> bool
{
    let mut start_byte = 0;
    let mut end_byte = text.len();

    // Find first character that shouldn't be trimmed
    for (byte_pos, ch) in text.char_indices() {
        if !should_trim(ch) {
            start_byte = byte_pos;  // Start OF the character, not after
            break;
        }
    }

    // Find last character that shouldn't be trimmed
    for (byte_pos, ch) in text.char_indices().rev() {
        if !should_trim(ch) {
            end_byte = byte_pos + ch.len_utf8();  // End AFTER the character
            break;
        }
    }

    // Handle edge cases
    if start_byte >= end_byte {
        return "";
    }

    &text[start_byte..end_byte]
}


// Trim normal whitespace, but keep HTML entities and Unicode whitespace
pub fn trim_normal_whitespace<'a>(text: &'a str) -> &'a str {
    trim_with_callback(text, is_normal_whitespace)
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trim_normal_whitespace() {
        // Should trim normal spaces
        assert_eq!(trim_normal_whitespace("  hello  "), "hello");

        // Should keep &nbsp; (non-breaking space)
        assert_eq!(trim_normal_whitespace("\u{00A0}hello\u{00A0}"), "\u{00A0}hello\u{00A0}");

        // Mixed case
        assert_eq!(trim_normal_whitespace("  \u{00A0}hello\u{00A0}  "), "\u{00A0}hello\u{00A0}");

        // All normal whitespace
        assert_eq!(trim_normal_whitespace("   \t\n   "), "");

        // All significant whitespace  
        assert_eq!(trim_normal_whitespace("\u{00A0}\u{00A0}"), "\u{00A0}\u{00A0}");
    }
}

File: packages/compiler/src/utils/whitespace.ts
/**
 * Whitespace handling utilities for GT Babel plugin
 *
 * CRITICAL: This must match the Rust implementation exactly to ensure
 * consistent whitespace handling between SWC and Babel plugins.
 */

/**
 * Types of whitespace characters
 */
enum WhitespaceType {
  NormalSpace, // Regular space from typing
  Tab, // Tab character from indentation
  Newline, // Newline from text formatting
  HtmlEntity, // Decoded HTML entities like &nbsp;
  UnicodeWhitespace, // Other Unicode whitespace
}

/**
 * Classify a character as a specific type of whitespace
 */
function classifyWhitespaceChar(ch: string): WhitespaceType | null {
  if (ch.length !== 1) return null;

  switch (ch) {
    case ' ':
      return WhitespaceType.NormalSpace;
    case '\t':
      return WhitespaceType.Tab;
    case '\n':
    case '\r':
      return WhitespaceType.Newline;
    case '\u00A0': // &nbsp;
    case '\u00AD': // &shy;
    case '\u202F': // Narrow no-break space
    case '\u2060': // Word joiner
      return WhitespaceType.HtmlEntity;
    default:
      // Check if it's other Unicode whitespace
      if (/\s/.test(ch)) {
        return WhitespaceType.UnicodeWhitespace;
      }
      return null;
  }
}

/**
 * Check if text contains significant whitespace (HTML entities or Unicode whitespace)
 * that must be preserved during transformation
 */
export function hasSignificantWhitespace(text: string): boolean {
  for (const ch of text) {
    const type = classifyWhitespaceChar(ch);
    switch (type) {
      case WhitespaceType.NormalSpace:
      case WhitespaceType.Tab:
      case WhitespaceType.Newline:
        continue;
      case WhitespaceType.HtmlEntity:
      case WhitespaceType.UnicodeWhitespace:
        return true;
      case null:
        continue; // Not whitespace
    }
  }
  return false;
}

/**
 * Check if a character is "normal" whitespace (space, tab, newline)
 * that can be safely trimmed
 */
export function isNormalWhitespace(ch: string): boolean {
  if (ch.length !== 1) return false;

  const type = classifyWhitespaceChar(ch);
  switch (type) {
    case WhitespaceType.NormalSpace:
    case WhitespaceType.Tab:
    case WhitespaceType.Newline:
      return true;
    case WhitespaceType.HtmlEntity:
    case WhitespaceType.UnicodeWhitespace:
    case null:
      return false;
  }
}

/**
 * Generic trim function with callback to determine what should be trimmed
 */
function trimWithCallback(
  text: string,
  shouldTrim: (ch: string) => boolean
): string {
  if (text.length === 0) return '';

  let startIndex = text.length;
  let endIndex = 0;

  // Find first character that shouldn't be trimmed
  let currentIndex = 0;
  for (const ch of text) {
    if (!shouldTrim(ch)) {
      startIndex = currentIndex;
      break;
    }
    currentIndex++;
  }

  // Find last character that shouldn't be trimmed
  currentIndex = 0;
  for (const ch of text) {
    if (!shouldTrim(ch)) {
      endIndex = currentIndex + 1; // End AFTER the character
    }
    currentIndex++;
  }

  // Handle edge cases
  if (startIndex >= text.length || endIndex === 0 || startIndex >= endIndex) {
    return '';
  }

  return text.slice(startIndex, endIndex);
}

/**
 * Trim normal whitespace (space, tab, newline) but preserve HTML entities
 * and Unicode whitespace which are significant for translation
 */
export function trimNormalWhitespace(text: string): string {
  return trimWithCallback(text, isNormalWhitespace);
}
